# Лабораторная работа №2
## Выполнила студентка группы БВТ2002 Бортникова Марина
## Задание №1
Реализовать методы поиска в соответствии с заданием. Организовать генерацию начального набора случайных данных. Для всех вариантов добавить реализацию добавления, поиска и удаления элементов. Оценить время работы каждого алгоритма поиска и сравнить его со временем работы стандартной функции поиска, используемой в выбранном языке программирования.

### Генерация массива

In [4]:
from random import randint
mass = [randint(0, 100) for i in range(8)]
mass = sorted(mass)
print (mass)

[13, 19, 20, 46, 46, 55, 66, 96]


## Бинарный поиск   O(log(n))
### Поиск

In [83]:
import time
startTime = time.time()
def binary_search(array, key):
    print(array)
    lower_bound = 0
    upper_bound = len(array) - 1

    while lower_bound <= upper_bound:
        center = (upper_bound + lower_bound) // 2

        if array[center] > key:
            upper_bound = center - 1

        elif array[center] < key:
            lower_bound = center + 1

        elif array[center] == key:
            return center

    return -1, lower_bound

print(binary_search(mass,17))
endTime = time.time()
totalTime = endTime - startTime

print("Время:", round(endTime-startTime,6), 'seconds')

[3, 6, 7, 7, 41, 49, 53, 69]
(-1, 4)
Время: 0.000998 seconds


### Удаление

In [5]:
def binary_delete(array, key):
    element = (binary_search(array, key))
    #print(array)
    if isinstance(element, tuple):
        return array
    del array[element]
    return (array)
    
print(binary_delete(mass.copy(), 17))

[4, 14, 17, 23, 36, 78, 94, 100]
[4, 14, 23, 36, 78, 94, 100]


### Вставка

In [6]:
def binary_paste(array, key):
    element = (binary_search(array, key))
    print(array)
    if isinstance(element, tuple):
        array.insert(element[1], key)
    else:
        array.insert(element, key)
    return (array)
    
print(binary_paste(mass.copy(), 1))

[4, 14, 17, 23, 36, 78, 94, 100]
[4, 14, 17, 23, 36, 78, 94, 100]
[1, 4, 14, 17, 23, 36, 78, 94, 100]


## Бинарное дерево O(log(n))

### Вставка

In [7]:
class Node:
    def __init__(self, x, parent):
        self.left = None
        self.right = None
        self.data = x
        self.parent = parent
class Tree:

    def __init__(self):
        self.root = None

    def insert(self, x, node=None):
        if node is None:
            node = self.root
            if self.root is None:
                self.root = Node(x, None)
                return

        if x < node.data:
            if node.left is None:
                node.left = Node(x, node)
                return
            self.insert(x, node.left)
        else:
            if node.right is None:
                node.right = Node(x, node)
                return
            self.insert(x, node.right)

    def print(self, node=None):
        if node is None:
            node = self.root

        if node.left:
            self.print(node.left)
        print(node.data)

        if node.right:
            self.print(node.right)



T = Tree()
T.insert(5)
T.insert(1)
T.insert(8)
T.insert(16)
T.insert(4)
T.insert(3)
T.insert(7)
T.insert(8)
T.insert(9)

T.print()


1
3
4
5
7
8
8
9
16


### Поиск

In [84]:
import time
startTime = time.time()
class Node:
    def __init__(self, x, parent):
        self.left = None
        self.right = None
        self.data = x
        self.parent = parent
class Tree:

    def __init__(self):
        self.root = None

    def insert(self, x, node=None):
        if node is None:
            node = self.root
            if self.root is None:
                self.root = Node(x, None)
                return

        if x < node.data:
            if node.left is None:
                node.left = Node(x, node)
                return
            self.insert(x, node.left)
        else:
            if node.right is None:
                node.right = Node(x, node)
                return
            self.insert(x, node.right)

    def search(self, x, node=None):
        if node is None:
            node = self.root

        if node.data == x:
            return node

        if x < node.data:
            if node.left is None:
                return False
            return self.search(x, node.left)
        else:
            if node.right is None:
                return False
            return self.search(x, node.right)



T = Tree()
T.insert(5)
T.insert(1)
T.insert(8)
T.insert(16)
T.insert(4)
T.insert(3)
T.insert(7)
T.insert(8)
T.insert(9)


print (T.search(8))
print (T.search(2))

endTime = time.time()
totalTime = endTime - startTime
print("Время:", round(endTime-startTime,6), 'seconds')

False
Время: 0.001992 seconds


### Удаление

In [9]:
class Node:
    def __init__(self, x, parent):
        self.left = None
        self.right = None
        self.data = x
        self.parent = parent
class Tree:

    def __init__(self):
        self.root = None

    def insert(self, x, node=None):
        if node is None:
            node = self.root
            if self.root is None:
                self.root = Node(x, None)
                return

        if x < node.data:
            if node.left is None:
                node.left = Node(x, node)
                return
            self.insert(x, node.left)
        else:
            if node.right is None:
                node.right = Node(x, node)
                return
            self.insert(x, node.right)

    def search(self, x, node=None):
        if node is None:
            node = self.root

        if node.data == x:
            return node

        if x < node.data:
            if node.left is None:
                return False
            return self.search(x, node.left)
        else:
            if node.right is None:
                return False
            return self.search(x, node.right)

    def print(self, node=None):
        if node is None:
            node = self.root

        if node.left:
            self.print(node.left)
        print(node.data)

        if node.right:
            self.print(node.right)

    def change_children(self, node, new_node):
        if node.parent is None:
            if new_node:
                new_node.parent = None
            self.root = new_node
            if node.left is new_node:
                self.root.right = node.right
            else:
                self.root.left = node.left
            return
        if node.parent.left is node:
            if new_node:
                new_node.parent = node.parent
                if node.left is new_node:
                    new_node.right = node.right
                else:
                    new_node.left = node.left
            node.parent.left = new_node
        else:
            if new_node:
                if node.left is new_node:
                    new_node.right = node.right
                else:
                    new_node.left = node.left
                new_node.parent = node.parent
            node.parent.right = new_node


    def rebalance(self, node):
        if node.left:
            if node.left.right:
                node.data = node.left.right.data
                node.left.right.data = None
                self.rebalance(node.left.right)
                return
            self.change_children(node, node.left)
            return
        if node.right:
            self.change_children(node, node.right)
            return
        self.change_children(node, None)

    def delete(self, x):
        node = self.search(x)
        if not node:
            return
        node.data = None
        self.rebalance(node)


T = Tree()
T.insert(5)
T.insert(1)
T.insert(8)
T.insert(16)
T.insert(4)
T.insert(3)
T.insert(7)
T.insert(8)
T.insert(9)

T.delete(5)
T.delete(8)

T.print()

1
3
4
7
8
9
16


## Фибоначчи O(ln(n))

### Поиск

In [77]:
import time
startTime = time.time()   
class FibonacciSearch:
    def __init__(self):
        self.i = 0
        self.p = 0
        self.q = 0
        self.stop = False

    def get_fibonacci_number(self, k):
        first_number = 0
        second_number = 1
        n = 0
        while n<k:
            temp_number = second_number
            second_number = first_number + second_number
            first_number = temp_number
            n = n+1
        return first_number

    def start_init(self, massiv):
        self.stop = False
        k = 0
        n = len(massiv)
        while (self.get_fibonacci_number(k+1) < len(massiv)):
            k = k+1
        m = self.get_fibonacci_number(k+1)-(n+1)
        self.i = self.get_fibonacci_number(k) - m
        self.p = self.get_fibonacci_number(k-1)
        self.q = self.get_fibonacci_number(k-2)

    def up_index(self):
        if self.p == 1:
            self.stop = True
        self.i = self.i + self.q
        self.p = self.p - self.q
        self.q = self.q - self.p

    def down_index(self):
        if self.q == 0:
            self.stop = True
        self.i = self.i - self.q
        temp = self.q
        self.q = self.p - self.q
        self.p = temp

    def search(self, massiv, element):
        self.start_init(massiv)
        result_index = -1
        while not self.stop:
            if self.i < 0:
                self.up_index()
            elif self.i >= len(massiv):
                self.down_index()
            elif massiv[self.i] == element:
                result_index = self.i
                break
            elif element < massiv[self.i]:
                self.down_index()
            elif element > massiv[self.i]:
                self.up_index()
        if result_index == -1:
            return result_index, self.i
        return result_index

fs = FibonacciSearch()
print (mass)
print(fs.search(mass,17))

endTime = time.time()
totalTime = endTime - startTime
print("Время:", round(endTime-startTime,6), 'seconds')

[9, 15, 31, 45, 57, 66, 66, 91]
(-1, 2)
Время: 0.001004 seconds


## Удаление

In [22]:
def fibinacci_delete(array, key):
    fs = FibonacciSearch()
    element = (fs.search(array,key))
    print(array)
    if isinstance(element, tuple):
        return array
    del array[element]
    return (array)
    
print(fibinacci_delete(mass.copy(), 17))

[4, 14, 17, 23, 36, 78, 94, 100]
[4, 14, 23, 36, 78, 94, 100]


### Вставка

In [23]:
def fibonacci_paste(array, key):
    fs = FibonacciSearch()
    element = (fs.search(array, key))
    print(array)
    if isinstance(element, tuple):
        array.insert(element[1], key)
    else:
        array.insert(element, key)
    return (array)
    
print(fibonacci_paste(mass.copy(), 24))

[4, 14, 17, 23, 36, 78, 94, 100]
[4, 14, 17, 23, 24, 36, 78, 94, 100]


## Интерполяционный
### Поиск

In [78]:
import time
startTime = time.time()
def interpolation_search(massiv, key):
    print (massiv)
    low = 0
    high = len(massiv)-1
    
    while (low<=high and key >= massiv[low] and key<= massiv[high]):
        middle = low + int(((high-low)/(massiv[high]-massiv[low]))*(key-massiv[low]))
        if massiv[middle] == key:
            return middle
        if massiv[middle] < key:
            low = middle + 1
        else:
            high = middle - 1
    return -1, high

print(interpolation_search(mass,17))

endTime = time.time()
totalTime = endTime - startTime
print("Время:", round(endTime-startTime,6), 'seconds')

[9, 15, 31, 45, 57, 66, 66, 91]
(-1, 7)
Время: 0.001992 seconds


### Вставка

In [35]:
def interpolation_paste(array, key):
    element = (interpolation_search(array, key))
    if isinstance(element, tuple):
        array.insert(element[1], key)
    else:
        array.insert(element, key)
    return (array)
    
print(interpolation_paste(mass.copy(), 96))

[4, 14, 17, 23, 36, 78, 94, 100]
[4, 14, 17, 23, 36, 78, 94, 96, 100]


### Удаление

In [36]:
def interpolation_delete(array, key):
    element = (interpolation_search(array,key))
    if isinstance(element, tuple):
        return array
    del array[element]
    return (array)
    
print(interpolation_delete(mass.copy(), 5))

[4, 14, 17, 23, 36, 78, 94, 100]
[4, 14, 17, 23, 36, 78, 94, 100]


## Встроенны поиск

In [79]:
import time
startTime = time.time()
def standart_search(massiv, element):
        return massiv.count(element)
    
print(standart_search(mass, 14))

endTime = time.time()
totalTime = endTime - startTime
print("Время:", round(endTime-startTime,6), 'seconds')

0
Время: 0.001 seconds


## Результаты

### Время выполнения бинарного поиска: 0.000999 seconds
### Время выполнения поиска в динарном дереве: 0.002994 seconds
### Время выполнения поиска Фибоначчи: 0.001004 seconds
### Время выполнения интерполиционного поиска: 0.001992 seconds
### Время выполнения встроенного потска: 0.001 seconds

### Вывод: Самым быстрым оказался встроенный поиск, самым медленным - поиск в бинарном дереве


# Задание 2 
## Простое рехеширование

In [47]:
class Hashtable:
    def __init__(self, size):
        self.data = [None] * size
        self.size = size

    def hash_function(self, value):
        return value % self.size

    def put(self, value):
        hash_value = self.hash_function(value)
        if self.data[hash_value] == value:
            return
        if self.data[hash_value] is None:
            self.data[hash_value] = value
        else:
            while self.data[hash_value] != None:
                hash_value = self.hash_function(hash_value + 1)
                if self.data[hash_value] == value:
                    return
                if self.data[hash_value] is None:
                    self.data[hash_value] = value
                    break

    def get(self, value):
        hash_value = self.hash_function(value)
        if self.data[hash_value] == value:
            return hash_value
        else:
            for i in range(self.size):
                hash_value = self.hash_function(hash_value + 1)
                if self.data[hash_value] == value:
                    return hash_value
        return False

    def delete(self, value):
        if not self.get(value):
            return -1
        hash_value = self.hash_function(value)
        if self.data[hash_value] == value:
            self.data[hash_value] = None
        else:
            for i in range(self.size):
                hash_value = self.hash_function(hash_value + 1)
                if self.data[hash_value] == value:
                    self.data[hash_value] = None

print (mass)                    
n = 10
h = Hashtable(n)
for i in mass:
    h.put(i)
print(h.data)

# Вставка
print("Вставка")
h.put(77)
print(h.data)

# Поиск
print("Поиск")
print (h.get(77))
print (h.get(19))

# Удаление
print("Удаление")
h.delete(77)
print(h.data)


[4, 14, 17, 23, 36, 78, 94, 100]
[100, None, None, 23, 4, 14, 36, 17, 78, 94]
Вставка
[100, 77, None, 23, 4, 14, 36, 17, 78, 94]
Поиск
1
False
Удаление
[100, None, None, 23, 4, 14, 36, 17, 78, 94]


## Рехеширование с помощью псевдослучайных чисел


In [50]:
import random
class Hashtable:
    def __init__(self, size):
        self.data = [None] * size
        self.size = size

    def hash_function(self, value):
        return value % self.size

    def put(self, value):
        hash_value = self.hash_function(value)
        if self.data[hash_value] == value:
            return
        if self.data[hash_value] is None:
            self.data[hash_value] = value
        else:
            random.seed(hash_value)
            while True:
                hash_value = self.hash_function(hash_value + random.randint(0, self.size))
                if self.data[hash_value] == value:
                    return
                if self.data[hash_value] is None:
                    self.data[hash_value] = value
                    break

    def get(self, value):
        hash_value = self.hash_function(value)
        if self.data[hash_value] == value:
            return hash_value
        else:
            random.seed(hash_value)
            for i in range(self.size):
                hash_value = self.hash_function(hash_value + random.randint(0, self.size))
                if self.data[hash_value] == value:
                    return hash_value
        return False



    def delete(self, value):
        if not self.get(value):
            return -1
        hash_value = self.hash_function(value)
        if self.data[hash_value] == value:
            self.data[hash_value] = None
        else:
            random.seed(hash_value)
            for i in range(self.size):
                hash_value = self.hash_function(hash_value + random.randint(0, self.size))
                if self.data[hash_value] == value:
                    self.data[hash_value] = None




print (mass) 
n = 10
h = Hashtable(n)
for i in mass:
    h.put(i)
print(h.data)

# Вставка
print("Вставка")
h.put(77)
print(h.data)

# Поиск
print("Поиск")
print (h.get(77))
print (h.get(19))

# Удаление
print("Удаление")
h.delete(77)
print(h.data)

[4, 14, 17, 23, 36, 78, 94, 100]
[100, 94, 17, 23, 4, None, 36, 14, 78, None]
Вставка
[100, 94, 17, 23, 4, None, 36, 14, 78, 77]
Поиск
9
False
Удаление
[100, 94, 17, 23, 4, None, 36, 14, 78, None]


## Метод цепочек

In [25]:
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None


class Hashtable:
    def __init__(self, size):
        self.data = []
        self.size = size
        for i in range (self.size):
            self.data.append(Node())

    def hash_function(self, value):
        return value % self.size

    def put(self, value):
        hash_value = self.hash_function(value)
        if self.data[hash_value].data is None:
            self.data[hash_value].data = value
        else:
            p = self.data[hash_value]
            while p.next is not None:
                p = p.next
            p.next = Node(value)

    def get(self, value):
        hash_value = self.hash_function(value)
        if self.data[hash_value].data == value:
            return True
        else:
            p = self.data[hash_value]
            while p is not None and p.data != value:
                p = p.next
            if p is not None and p.data == value:
                return True
        return False

    def delete(self, value):
        if not self.get(value):
            return False
        hash_value = self.hash_function(value)
        if self.data[hash_value].data == value:
            self.data[hash_value].data = None
        else:
            p = self.data[hash_value]
            pre = None
            while p is not None and p.data != value:
                pre = p
                p = p.next
            if p is None:
                return False
            else:
                pre.next = p.next

    def print (self):
        for i in h.data:
            if not h.data:
                continue
            print(i.data)
            t = i
            while t.next:
                print(t.next.data)
                t = t.next


print (mass)               
n = 10
h = Hashtable(n)
for i in mass:
    h.put(i)

h.print()

#Вставка
print(" ")
print("Вставка")
h.put(77)
h.print()

# Поиск
print(" ")
print("Поиск")
print (h.get(77))
print (h.get(19))

# Удаление
print(" ")
print("Удаление")
h.delete(77)
h.print()

[13, 19, 20, 46, 46, 55, 66, 96]
20
None
None
13
None
55
46
46
66
96
None
None
19
 
Вставка
20
None
None
13
None
55
46
46
66
96
77
None
19
 
Поиск
True
True
 
Удаление
20
None
None
13
None
55
46
46
66
96
None
None
19


# Задание 3
Расставить на стандартной 64-клеточной шахматной доске 8 ферзей так, чтобы ни один из них не находился под боем другого». Подразумевается, что ферзь бьёт все клетки, расположенные по вертикалям, горизонталям и обеим диагоналям Написать программу, которая находит хотя бы один способ решения задач.

In [86]:
global queen
global number
queen = [None] * 8
number = 0


def print_table():
    global number
    number += 1
    print('\n%d решение \t' % number)
    for x in range(8):
        for y in range(8):
            if x == queen[y]:
                print('<q>', end=' ')
            else:
                print('<->', end=' ')
        print('\t')


def attack(row, col):
    global queen
    for i in range(col):
        offset_col = abs(i - col)
        offset_row = abs(queen[i] - row)
        if queen[i] == row or offset_row == offset_col:
            return True
    return False


def decide_position(value):
    global queen
    for i in range(8):
        if not attack(i, value):
            queen[value] = i
            if value == 7:
                print_table()
            else:
                decide_position(value + 1)


decide_position(0)


1 решение 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <q> <-> <-> <-> <-> <-> 	

2 решение 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <-> <q> <-> <-> <-> <-> <-> 	

3 решение 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <-> <q> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	

4 решение 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <-> <-> <-> <-> <q> <-> <-

<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <q> <-> <-> <-> <-> <-> 	

49 решение 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <q> <-> <-> <-> <-> <-> 	

50 решение 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <q> <-> <-> <-> <-> <-> 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	

51 решение 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <-> <q> <-> <-> <-> <-> <-> 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
<-> <-

<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<q> <-> <-> <-> <-> <-> <-> <-> 	

92 решение 	
<-> <-> <q> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <q> <-> <-> 	
<-> <-> <-> <q> <-> <-> <-> <-> 	
<-> <q> <-> <-> <-> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <-> <q> 	
<-> <-> <-> <-> <q> <-> <-> <-> 	
<-> <-> <-> <-> <-> <-> <q> <-> 	
<q> <-> <-> <-> <-> <-> <-> <-> 	
